Separate build and test logs

apjanke opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally? _No; I'm getting a warning that needs to be cleaned up.

-----

Any objections to splitting the build and test logs in to separate subdirectories?

Currently, `brew install` and `brew test` both drop their log files in the same `.../Logs/Homebrew/<formula_name>` directory. This means that if you do a `brew install foo` and then  `brew test foo` on a formula that includes `system()` calls in its `def test`, its test logs will contaminate the installation logs. This happens with `postgresql`, for example.

I suspect this was unintentional, and just wasn't noticed because we seldom use Homebrew logs except for immediately after a failed `brew install`, in which case a `brew test` will not have been run.

This PR adds new `build/` and `test/` subdirs to separate the logs for the two stages, and adjusts `brew gist-logs` and other things using logs to account for it. `brew cleanup` will handle migrating logs in the old directory structure to the new one, so users won't lose existing logs.

This has the additional advantage of opening up logical space in the `brew` logs directory to allow for logging of other things, if that becomes useful in the future.

###   Failing tests   ###

The `brew tests` run isn't failing outright, but it is complaining about leaked files. I haven't been able to fix this yet, because I can't figure out where the whitelist for "okay to appear" files or log structure skeleton is, or if it's complaining about this for some other reason. Help on this is welcome.

Marking this as "in progress" because of this, and won't ask to merge until it's clean.

(The exact test that appears to leak will vary across `brew tests` runs, I think because the random seed shuffles the order tests are run in, but `test_cleanup_logs` is always involved.)

```
[✘ /usr/local/Library/Homebrew on ⇄ separate-build-and-test-logs]
$ brew tests
Run options: --seed 5148

# Running:

.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 51.091141s, 16.7348 runs/s, 40.5941 assertions/s.

855 runs, 2074 assertions, 0 failures, 0 errors, 0 skips
Warning: File leak is detected
InstallTests#test_a_basic_install
--- expected
+++ actual
@@ -3,6 +3,7 @@
  "/cellar",
  "/formula_cache",
  "/logs",
+ "/logs/build",
  "/prefix",
  "/prefix/Library",
  "/prefix/Library/Taps",
CleanupTests#test_cleanup_logs
--- expected
+++ actual
@@ -3,7 +3,6 @@
  "/cellar",
  "/formula_cache",
  "/logs",
- "/logs/build",
  "/prefix",
  "/prefix/Library",
  "/prefix/Library/Taps",
```

###   Considerations   ###

This will be a breaking change for any external commands or other code that relies on the current directory structure of the log files.

xu-cheng commented:
1. May I suggest to make the log structure like this: `HOMEBREW_LOGS/<formula>/{build,postinstall,test}`? In this way, it is easier to view the log from the same formula. And make `brew build-logs` happy, which I used a lot and is available from `homebrew/dev-tools`.
2. You need to handle post_install.
2. Instead of adding more methods to `formula`, can we simply set the log path directly in `build.rb/postinstall.rb/test.rb`. I hope we can achieve the same goal without adding too much complexity

apjanke commented:
Good point; I had overlooked `post_install`. Thanks. I've amended this PR with an attempt to address this.

1. Yeah, that probably makes more sense. Keeps related logs closer together, and less chance of the change breaking things that use `Formula.logs`.

How about `HOMEBREW_LOGS/Formula/<formula>/{install,post_install,test,...}`? This still keeps all the per-formula logs together, and plain `Formula.logs()` meaningful, but the extra `Formula` level of indirection leaves space for logging other things if needed. It wouldn't be used right away, but since this directory structure reorganization is a potentially breaking change, I'd like to create the ability to add other logging stuff later without requiring an additional breaking change.

This naming scheme uses `install`/`post_install`/`test` so the stage names exactly match the method names of the stages in formulae. Though maybe it should be `postinstall` instead, to match the subcommand names instead of the DSL methods.

I've changed this PR to do this. Under it, `brew build-logs` keeps working without modification. For example, here's `brew build-logs anjuta` after a `brew install -s anjuta`.

<img width="802" alt="brew build-logs anjuta under separate build and test log dirs" src="https://cloud.githubusercontent.com/assets/2618447/15626192/323a28d2-248c-11e6-832b-6b43b67a812a.png">

I think this not only works, but is an improvement, because now the post_install step is broken out in to a separate group. Here's the current `brew build-logs anjuta` for comparison; note the duplication in steps 01 and 02.

<img width="837" alt="brew build-logs current for anjuta" src="https://cloud.githubusercontent.com/assets/2618447/15626470/67055b70-2493-11e6-9eb8-ae23fb390c92.png">

(The new version happens to be ordered correctly in `build-logs`, too, but that's just an accident of alphabetical ordering. I don't think we can enforce ordering here without introducing another place for potentially breaking changes upon future reorganizations where an additional step is added. Maybe that's worth it, but my first thought is no.)

2. Handled post_install.

3. This other way seems more complex to me, because then control of the formula-related log directory structure is distributed throughout the codebase. It'd effectively be using a public mutable property to control the internal behavior of `Formula` code, since it's `Formula.system` that's writing the logs. So someone not familiar with the code and wanting to learn the log structure would have to read the entire code base to determine all the places where the log directory might be set. The way this PR does it now, all the log dir structure code is in one file, `formula.rb`, so you can read that and be confident you get the whole thing. (To the extent that you can do that in a monkey-patchable language like Ruby.)

The `Formula` class is getting quite large, but in my view, adding these few methods just linearly adds to that complexity, whereas opening it up for setting log paths and distributing that logic throughout the codebase adds another degree of freedom, and no validity constraints on it, which is a greater increase in overall codebase complexity IMHO.

If the size of `Formula` is a concern, maybe we could factor out the logging-related code to a separate `Logs` class. I feel like the excessive size of `Formula` here is a consequence of its existing design, and `Formula` (or a dedicated auxiliary class) is still the right place for this logic.

####  Testing

I tested this with `brew install -s anjuta`, since it has a `post_install` step with `system` calls in it, and so do some of its dependencies.

####  Outstanding TODOs

* `gist-logs` does not pick up the post_install logs as part of the build logs, but I think it probably should.
* The `sandbox.build.log` should probably be relocated to `install/sandbox.build.log`, and there should probably be support for a `sandbox.postinstall.log`.

chdiza commented:
I think it would be better to leave the build logs "loose" in `$HOMEBREW_LOG/formula_name`, and put the test logs in `$HOMEBREW_LOG/formula_name/test`.  (Assuming this still fixes the problem this PR was created to solve.)

The fewer levels of navigation necessary, the better.

Also, it wouldn't break anybody's scripts that might depend on build logs being where they currently are.

apjanke commented:
Hmm. I personally like my current arrangement better, but I think you're probably right here.

> I think it would be better to leave the build logs "loose" in `$HOMEBREW_LOG/formula_name`, and put the test logs in `$HOMEBREW_LOG/formula_name/test`.  (Assuming this still fixes the problem this PR was created to solve.)

Yes, that would solve the problem, I think; it provides the needed separation. (Along with a `HOMEBREW_LOGS/<formula_name>/postinstall`.) And it does seem less prone to break things. Especially because the state after a failed build, which is by far the most common time to use logs (I think), would be the same as the current log handling logic, and this would only change how logs are arranged after a successful build/test.

I think it also has the advantage of ensuring that the `test` and `postinstall` logs for a formula correspond to the version installed by `brew install` whose logs are there. In my current implementation, if you do `brew install -s foo`, `brew test foo`, and then later `brew upgrade foo`, you could end up with a `test` log from an old version of `foo` and the `install` log from the newer one. Or if installation failed during the main installation, you could have a leftover `postinstall` step from a previous installation. Those seem undesirable.

>  The fewer levels of navigation necessary, the better.

Hmm. I'd rather get rid of some of the need for manual navigation by introducing a `brew info --log-path` option or similar, so users could write up a little shell alias and jump right to the appropriate logs, instead of having to manually do the `cd ~/Library/Logs/Homebrew` part or know the exact log dir structure. And non-Ruby external scripts could be switched to use it, so they no longer have to embed knowledge of the log directory structure.

Navigating within these per-formula subdirectories would be convenient, since `postinstall` and `test` would naturally sort after all the numbered files from the build.

> Also, it wouldn't break anybody's scripts that might depend on build logs being where they currently are.

It could still break some things: anything that depends on the current case of every child of the formula log directory being a regular file and a log file could choke once we introduce subdirectories. (E.g. `brew gist-logs`.) But those are probably a minority of scripts, and could be fixed with small tweaks. And anything run after a failed build would have no chance of breakage (unless the failure was specifically in the `postinstall` stage) because the dir layout is unchanged at that point.

My gut really likes the tidiness of `HOMEBREW_LOGS/Formula/<name>/{install,postinstall,test}`, but I have to admit that your version is probably better for users (both end users and contributors), and we could always add in the `Formula` indirection later if & when the need actually arises.

Any other folks see advantages for one way or the other? I haven't tried out the actual code changes yet, but chdiza's way does seem a lot more conservative, and would probably be a non-breaking change for most uses outside the core `brew` code.

xu-cheng commented on Library/Homebrew/cleanup.rb:
```diff
@@ -33,7 +35,15 @@ def self.cleanup_formula(formula)
       return unless HOMEBREW_LOGS.directory?
       HOMEBREW_LOGS.subdirs.each do |dir|
-        cleanup_path(dir) { dir.rmtree } if prune?(dir, :days_default => 14)
+        if Formulary.formula_ref? dir.basename.to_s
```
Can we avoid this? I'm kinda down to add new method to Formualry.

apjanke replied:
We could switch it to detect formulae the other way, by excluding the names known to be installation steps ("install", "postinstall", "test") and assuming everything else is a formula.

Come to think of it, that may be more correct anyway, because formulae may have been removed or renamed over the course of a `brew update` since the installation was initially done.

It would accidentally move the logs from any external commands which are dropping their own log files under `HOMEBREW_LOGS`, but that seems like a small risk with low impact.

MikeMcQuaid replied:
> We could switch it to detect formulae the other way, by excluding the names known to be installation steps ("install", "postinstall", "test") and assuming everything else is a formula.

👍

xu-cheng commented on Library/Homebrew/formula.rb:
```diff
@@ -818,8 +850,10 @@ def post_install_defined?
-    build, self.build = self.build, Tab.for_formula(self)
-    post_install
+    with_logging(:post_install) do
+      build, self.build = self.build, Tab.for_formula(self)
```
It looks weird to have this line inside block, especially considering the ensure block

apjanke replied:
Moved outside the block.

xu-cheng commented on Library/Homebrew/formula.rb:
```diff
@@ -690,10 +695,37 @@ def bottle_prefix
-    HOMEBREW_LOGS+name
+  def logs(log_type = :main)
+    case log_type
+    when :main
```
The main is a little confusing.

xu-cheng replied:
And I think it should be avoid, you can always pass the type in the sandbox rules, which is the only place it's used.

apjanke replied:
I'm not sure that works, because the per-step subdirectories are created conditionally inside the sandboxed processes only as needed, and I think that counts as a modification to the parent directory. I'll test it to be sure.

Would a different name help?

I think it's useful to be able to get the path to the top log directory for a given formula, e.g. for code that doesn't know or care what all steps might exist under it, like if you wanted to `cd` the user to it for manual examination, or just create a tarball of everything under it.

xu-cheng replied:
Nope, the mkpath for log is happening outside of sandbox.

UniqMartin commented:
I haven't had a chance to review the code changes yet, but may I suggest a somewhat simpler and less invasive approach? My suggestion would be to leave the log files in the directory where they are currently placed and just change the code such that the files get an additional prefix that signifies the corresponding stage:

- `install.01.<command>` etc. for the formula `install` stage. Could be also `build.01.<command>` or even stay as `01.<command>` as these are the files most people care about and logs from the other stages are less likely to even be created. (Keeping the old name has the benefit that nobody will be surprised by a rename/move while still keeping things tidy.)
- `postinstall.01.<command>` etc. for the `postinstall` stage.
- `test.01.<command>` etc. for the `test` stage.

How does that sound?

UniqMartin commented:
Forgot to mention this in the previous comment: :+1: for tackling this issue. It has frequently annoyed me that these logs can overlap, as it's not too uncommon for me to run all three steps locally, but apparently not enough to pursue a change. (Sorting the directory listing by time is my current workaround.)

MikeMcQuaid commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -7,6 +7,9 @@
 #:     GitHub repository as well as creating the Gist.
 #:
+#:     If `--test-logs` is passed, uploads the logs for the formula's last test run, instead
+#:     of the build logs.
```
Could maybe always upload them all? Either way, not sure this makes sense to be an end-user option.

MikeMcQuaid replied:
Also maybe worth thinking about `postinstall`

apjanke replied:
Yeah. After seeing other folks' feedback about alternate dir organization, I think "just always upload everything" is the right behavior, too. Going to remove this option and fix `gist-logs` to include the logs from all steps, with some rename logic to make it sensible.

MikeMcQuaid commented on Library/Homebrew/build.rb:
```diff
@@ -128,7 +128,7 @@ def install
         formula.prefix.mkpath
 
-        formula.install
+        formula.run_install
```
Would be nice to avoid renaming this if possible.

apjanke replied:
It's possible, but ugly, with this general approach: like `run_test`, `install` is overridden by the DSL definitions, so we can't modify it directly, but we need a level of indirection to introduce the logging setup before calling it directly. We could avoid a it by adding that `with_logging` call at this point in `build.rb`, but I think that's the wrong place, because it should _always_ be called when `install` will be run, not just when it's done through this `build.rb` file. The way to avoid introducing a `run_install` is to make an implicit rule that "every place that calls `install` needs to wrap it with a `with_logging do` call", which seems worse than introducing a new method that does that. Introducing a `run_install` level formalizes that and puts it in the code structure, instead of doco which contributors may or may not have read.

MikeMcQuaid replied:
Makes sense 👍

MikeMcQuaid commented:
@apjanke Could you get this fixed up and merged in? Thanks!

apjanke commented:
Okay, I've cleaned this up, addressed feedback, and rebased on master.

I switched to using a flat directory structure with prefixes on the log file names themselves, like Martin suggested. Did this because it:

* Results in simpler code changes.
* Avoids introducing extra Formula methods.
* Keeps files in same dir arrangement, like chdiza suggested.
* In the basic build case, keeps log file names exactly as before, reducing user-visible changes.

The "sandbox.*.log" files are renamed to "*.sandbox.log" so they now sort with the phase they are describing.

Here's what the log file names look like now:

```
[~/Library/Logs/Homebrew/anjuta]
$ ls
01.configure     02.make.cc         post_install.01.glib-compile-schemas    test.01.anjuta
01.configure.cc  build.sandbox.log  post_install.02.gtk3-update-icon-cache  test.sandbox.log
02.make          config.log         post_install.03.update-mime-database
```

`brew gist-logs` picks up all files for the formula. Here's an example gist-logs made using these changes: https://gist.github.com/a5ab0b16f1aab062d0abf041fb42e02a

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -730,12 +735,32 @@ def bottle_prefix
+  # The directory where the formula's installation or test logs will be written.
   # @private
-  def logs
+  def logs()
```
Why add the `()`?

apjanke replied:
Oops. Just an incomplete reversion of the prior change to `logs(log_type)`. Fixed.

MikeMcQuaid commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -1,4 +1,4 @@
-#:  * `gist-logs` [`--new-issue`|`-n`] <formula>:
+#:  * `gist-logs` [`--new-issue`|`-n`] [`--test-logs`] <formula>:
```
But what does `--test-logs` do? 😉

apjanke replied:
Fixed. :)

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -730,12 +735,32 @@ def bottle_prefix
   end
 
+  # The prefix, if any, to use in filenames for logging current activity
+  def active_log_prefix()
```
Another `()`

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -730,12 +735,32 @@ def bottle_prefix
 
+  # The prefix, if any, to use in filenames for logging current activity
+  def active_log_prefix()
+    if @active_log_type.nil?
```
I'd generally flip the test to be just for `@active_log_type` rather than `.nil?`

MikeMcQuaid replied:
Also: why use `@active_log_type` here and `active_log_type` below?

apjanke replied:
Old Java habits. Fixed.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -856,6 +881,16 @@ def post_install_defined?
 
+  # @private
+  def run_post_install
+    build, self.build = self.build, Tab.for_formula(self)
```
Split this into two assignments.

MikeMcQuaid commented:
A few more nits.

apjanke commented:
Nits picked, rebased on master. Anything else before merging?

MikeMcQuaid commented:
👍 :shipit:

UniqMartin commented on Library/Homebrew/cleanup.rb:
```diff
@@ -31,7 +31,6 @@ def self.cleanup_formula(formula)
     end
 
     def self.cleanup_logs
-      return unless HOMEBREW_LOGS.directory?
```
Is this indeed no longer needed? This looks like a leftover from the original implementation of this PR.

apjanke replied:
You're right; probably still needed. Restoring.

UniqMartin commented on Library/Homebrew/formula.rb:
```diff
@@ -127,6 +127,11 @@ class Formula
   attr_accessor :local_bottle_path
 
+  # When performing a build, test, or other loggable action, indicates which
+  # log file location to use.
```
Strictly speaking, this is now the log file prefix, not the log file location, right? Maybe it would even make sense to explicitly enumerate all the values this can be set to in this documentation comment.

apjanke replied:
Well, a file name is still a location, right? ;) And I still think of it is a "type" rather than just a prefix; the prefix is derived from it.

The values are unconstrained and unchecked now, and you could stick any (valid file name substring) value in there. Could enumerate the ones the rest of the code uses, but if we want to do that, maybe it should go back to being a constrained symbol type? It feels a little un-Rubyish to document a limit that's just the state of current usage by the rest of the code.

UniqMartin replied:
I'm definitely against overengineering this and prematurely constraining this to a fixed set of symbols. Feel free to leave everything as-is; and thanks for your thoughts!

UniqMartin commented on Library/Homebrew/formula.rb:
```diff
@@ -730,12 +735,32 @@ def bottle_prefix
     HOMEBREW_LOGS+name
   end
 
+  # The prefix, if any, to use in filenames for logging current activity
```
It would be nice if the documentation comments were always complete sentences, i.e. I'm missing a trailing full stop here and in a few other places that is present in (most?) other such comments.

apjanke replied:
Few of the method comments we have in existing code seem to be complete sentences. This seems more consistent with the existing style. (Though not many methods are have method-level comments, so it's not clear there's a consistent style.)

UniqMartin replied:
I don't feel strongly about this, thus feel free to leave this as-is.

This was just an idea to try and standardize a little bit on our (generally lacking) API documentation. If the description consists of complete sentences, it's fairly easy to add something to it. Otherwise I'm often wondering: Do I add a full stop now and continue with a new sentence? Do I delimit with a somewhat awkward semicolon? (And maybe I'm just overthinking such simple matters. 😀)

UniqMartin commented on Library/Homebrew/formula.rb:
```diff
@@ -730,12 +735,32 @@ def bottle_prefix
+    ensure
+      @active_log_type = old_log_type
+    end
+  end
```
I think you can simplify this to the following without any change in meaning (unless we can expect the first assignment to fail):

```ruby
def with_logging(log_type)
  old_log_type = @active_log_type
  @active_log_type = log_type
  yield
ensure
  @active_log_type = old_log_type
end
```

apjanke replied:
I think that's true, and it tested out when I tried it. Switched to that form. Look better?

UniqMartin commented on Library/Homebrew/formulary.rb:
```diff
@@ -276,6 +276,10 @@ def self.canonical_name(ref)
 
+  def self.formula_ref?(ref)
+    !loader_for(ref).is_a? NullLoader
+  end
```
Is this still needed? This looks like a leftover from the original implementation of this PR.

apjanke replied:
Yeah, leftover. Removing.

UniqMartin commented on Library/Homebrew/test.rb:
```diff
@@ -25,7 +25,9 @@
-    raise "test returned false" if formula.run_test == false
+    formula.with_logging("test") do
+      raise "test returned false" if formula.run_test == false
+    end
```
I'm a bit surprised there's a `Formula#with_logging` invocation both here and in the `Formula#run_test` method called from here. Which one of these is redundant? Or are they both needed and why?

apjanke replied:
They're redundant. Doesn't really matter which one, since they guard the same potential `system` calls. Removing this one in favor of the one inside `run_test`, for consistency with `run_postinstall`, and because it's possible `run_test` could get called from elsewhere, and it should probably establish that logging context regardless of the caller.

UniqMartin commented:
I pointed out a few more nits and a few a-bit-more-than-nits. 🙈

> I switched to using a flat directory structure with prefixes on the log file names themselves, like Martin suggested. Did this because it:

Awesome work! Thanks a lot! :bow:

apjanke commented:
Further nits addressed.

> Awesome work! Thanks a lot!

You're welcome!

UniqMartin commented:
:+1: Great work! Finally log files that don't overwrite each other! :tada:

apjanke commented:
Merged. Thanks for the help an patience, everyone!
